java/Mekhmet-Efe_Adzhar/src/com/zoolatech/lecture8/tasks/_1/Number1.java

```diff
@@ -0,0 +1,124 @@
+*/
+
+public class Number1 {
+    public static void main(String[] args) throws InterruptedException {
+        Lock lock = new ReentrantLock();
+
+        Restaurant restaurant = new Restaurant();
+        AtomicInteger atomicInteger = new AtomicInteger(1);
+        Order order = new Order(atomicInteger, "Efe");
+
+        BlockingQueue<String> orderQueue = new ArrayBlockingQueue<>(1);
+        BlockingQueue<String> ordersTaken = new ArrayBlockingQueue<>(1);
+
+        Thread customerQueue = new Thread(() -> {
+            while (!Thread.currentThread().isInterrupted()) {
+                try {
+                    orderQueue.put(restaurant.acceptOrder(order));
+                    Thread.sleep(2000);
+                    System.out.println("added order number :" + (order.getId()) + " " + Thread.currentThread().getName());
+                    Thread.sleep(2000);
+                } catch (InterruptedException e) {
+                    e.printStackTrace();
+                }
+            }
+        });
+
+        Thread chef = new Thread(() -> {
+            while (!Thread.currentThread().isInterrupted()) {
+                try {
+                    String currentOrder = orderQueue.take();
+                    Thread.sleep(2000);
+                    System.out.println("Cooking order: " + currentOrder + " " + Thread.currentThread().getName());
+                    Thread.sleep(2000);
+                    ordersTaken.put(currentOrder);
+                    System.out.println("Preparing order to delivery person: " + currentOrder + " " + Thread.currentThread().getName());
+                    Thread.sleep(2000);
+                } catch (InterruptedException e) {
+                    e.printStackTrace();
+                }
+            }
+        });
+
+        Thread deliveryPerson = new Thread(() -> {
+            while (!Thread.currentThread().isInterrupted()) {
+                try {
+                    Thread.sleep(2000);
+                    System.out.println("Picked up and taking order back : " + ordersTaken.take() + " " + Thread.currentThread().getName());
+                    Thread.sleep(2000);
+                } catch (InterruptedException e) {
+                    e.printStackTrace();
+                } finally {
+                    order.getOrder();
+                }
+            }
+        });
+
+        customerQueue.setName("CUSTOMER");
+        chef.setName("CHEF");
+        deliveryPerson.setName("DELIVERY PERSON");
+
+        customerQueue.start();
+        chef.start();
+        deliveryPerson.start();
+
+    }
+}
+
+class Restaurant {
+    Lock lock = new ReentrantLock();
+
+    public String acceptOrder(Order order) {
+        lock.lock();
+        try {
+            return order.getId() + " " + order.getCustomerName();
+        } finally {
+            lock.unlock();
+        }
+    }
+}
+
+class Order {
+    AtomicInteger id;
+    volatile String customerName;
```

Comment:
1. `customerName` isn't modified from other threads. So, we don't need to define it as `volatile`. 
2. `customerName` doesn't change after object creation, so it can be defined as final
3. all instance variables can be defined as private
4. id can be just an `int` field - it doesn't change after object creation (same it should be private and final)
5. since getOrder works with immutable data - explicit lock can be removed